planner/core/rule_max_min_eliminate.go

```diff
@@ -18,71 +18,188 @@ import (
-			if !mysql.HasNotNullFlag(f.Args[0].GetType().Flag) {
-				sel := LogicalSelection{}.Init(ctx)
-				isNullFunc := expression.NewFunctionInternal(ctx, ast.IsNull, types.NewFieldType(mysql.TypeTiny), f.Args[0])
-				notNullFunc := expression.NewFunctionInternal(ctx, ast.UnaryNot, types.NewFieldType(mysql.TypeTiny), isNullFunc)
-				sel.Conditions = []expression.Expression{notNullFunc}
-				sel.SetChildren(p.Children()[0])
-				child = sel
-			}
+// cloneSubPlans clones the subPlan. We only consider `Selection` and `DataSource` here,
+// because we have restricted the subPlan in `checkColCanUseIndex`.
+func (a *maxMinEliminator) cloneSubPlans(plan LogicalPlan) LogicalPlan {
+	switch p := plan.(type) {
+	case *LogicalSelection:
+		conditions := make([]expression.Expression, 0, len(p.Conditions))
+		for _, cond := range p.Conditions {
+			conditions = append(conditions, cond.Clone())
+		}
+		sel := LogicalSelection{Conditions: conditions}.Init(p.ctx)
+		sel.SetChildren(a.cloneSubPlans(p.children[0]))
+		return sel
+	case *DataSource:
+		// This is a shallow clone which may not be safe.
+		newDs := *p
+		newDs.self = &newDs
+		return &newDs
+	}
+	// This won't happen, because we have checked the subtree.
+	return nil
+}
 
-			// Add Sort and Limit operators.
-			// For max function, the sort order should be desc.
-			desc := f.Name == ast.AggFuncMax
-			// Compose Sort operator.
-			sort := LogicalSort{}.Init(ctx)
-			sort.ByItems = append(sort.ByItems, &ByItems{f.Args[0], desc})
-			sort.SetChildren(child)
-			child = sort
+// splitAggFuncAndCheckIndices splits the agg to multiple aggs and check whether each agg needs a sort
+// after the transformation. For example, we firstly split the sql: `select max(a), min(a), max(b) from t` ->
+// `select max(a) from t` + `select min(a) from t` + `select max(b) from t`.
+// Then we check whether `a` and `b` have indices. If any of the used column has no index, we cannot eliminate
+// this aggregation.
+func (a *maxMinEliminator) splitAggFuncAndCheckIndices(agg *LogicalAggregation) (aggs []*LogicalAggregation, canEliminate bool) {
+	aggs = make([]*LogicalAggregation, 0, len(agg.AggFuncs))
+	for _, f := range agg.AggFuncs {
+		// We must make sure the args of max/min is a simple single column.
+		col, ok := f.Args[0].(*expression.Column)
+		if !ok {
+			return nil, false
 		}
+		if !a.checkColCanUseIndex(agg.children[0], col) {
+			return nil, false
+		}
+	}
+	// we can split the aggregation only if all of the aggFuncs pass the check.
+	for i, f := range agg.AggFuncs {
+		newAgg := LogicalAggregation{AggFuncs: []*aggregation.AggFuncDesc{f}}.Init(agg.ctx)
+		newAgg.SetChildren(a.cloneSubPlans(agg.children[0]))
+		newAgg.schema = expression.NewSchema(agg.schema.Columns[i])
+		aggs = append(aggs, newAgg)
+	}
+	return aggs, true
+}
 
-		// Compose Limit operator.
-		li := LogicalLimit{Count: 1}.Init(ctx)
-		li.SetChildren(child)
+// Try to convert a single max/min to Limit+Sort operators.
```

Comment:
In golang, comment for a function is recommended to be started with the function name. For example, the comment for this function should be started with `eliminateSingleMaxMin`.